README.md
# gatsby-remark-autolink-headers-turkish

Adds GitHub-style hover links to headers in your markdown files when they're rendered.

This is a sub-plugin for `gatsby-transformer-remark`. As demoed below, add this plugin to the options of `gatsby-transformer-remark`.

## Install

`npm install gatsby-remark-autolink-headers-turkish`

## How to use

```javascript
// In your gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [`gatsby-remark-autolink-headers-turkish`],
      },
    },
  ],
}
```

Note: if you are using `gatsby-remark-prismjs`, make sure that it’s listed after this plugin. Otherwise, you might face an issue described here: https://github.com/gatsbyjs/gatsby/issues/5764.

```javascript
// good
{
  resolve: `gatsby-transformer-remark`,
  options: {
    plugins: [
      `gatsby-remark-autolink-headers-turkish`,
      `gatsby-remark-prismjs`,
    ],
  },
}

// bad
{
  resolve: `gatsby-transformer-remark`,
  options: {
    plugins: [
      `gatsby-remark-prismjs`, // should be placed after `gatsby-remark-autolink-headers`
      `gatsby-remark-autolink-headers-turkish`,
    ],
  },
}
```

## Options

- `offsetY`: Signed integer. Vertical offset value in pixels (optional)
- `icon`: SVG shape inside a template literal or boolean `false`. Set your own svg or disable icon (optional)
- `className`: String. Set your own class for the anchor (optional)
- `maintainCase`: Boolean. Maintains the case for markdown header (optional)
- `removeAccents`: Boolean. Remove accents from generated headings IDs (optional)
- `enableCustomId`: Boolean. Enable custom header IDs with `{#id}` (optional)
- `isIconAfterHeader`: Boolean. Enable the anchor icon to be inline at the end of the header text (optional)
- `elements`: String array. Specify which type of header tags to link (optional)

```javascript
// In your gatsby-config.js
module.exports = {
  plugins: [
    {
      resolve: `gatsby-transformer-remark`,
      options: {
        plugins: [
          {
            resolve: `gatsby-remark-autolink-headers-turkish`,
            options: {
              offsetY: `100`,
              icon: `<svg aria-hidden="true" height="20" version="1.1" viewBox="0 0 16 16" width="20"><path fill-rule="evenodd" d="M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z"></path></svg>`,
              className: `custom-class`,
              maintainCase: true,
              removeAccents: true,
              isIconAfterHeader: true,
              elements: [`h1`, `h4`],
            },
          },
        ],
      },
    },
  ],
}
```

## How to style the anchor link

By default, the anchor link has a class of `anchor` (see `className` option to change this name) on the element but has no additional styling. To make it fit your website, you'll have to write some CSS to change the appearance.

In your CSS you can specify this element, in this instance the anchor tag will appear red:

```css
a.anchor {
  fill: "red";
}
```

Note: There are a variety of approaches to styling your Gatsby site, see [styling documentation](https://www.gatsbyjs.com/docs/styling/) for more detail.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [1.0.0](https://github.com/gatsbyjs/gatsby/compare/gatsby-remark-autolink-headers@2.8.0-next.0...gatsby-remark-autolink-headers@2.8.0) (2020-12-15)

**Note:** Version bump only for package gatsby-remark-autolink-headers

gatsby-ssr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _react = _interopRequireDefault(require("react"));

var pluginDefaults = {
  className: "anchor",
  icon: true,
  offsetY: 0
};

exports.onRenderBody = function (_ref, pluginOptions) {
  var setHeadComponents = _ref.setHeadComponents;

  var _Object$assign = Object.assign(pluginDefaults, pluginOptions),
      className = _Object$assign.className,
      icon = _Object$assign.icon,
      offsetY = _Object$assign.offsetY;

  var styles = "\n    ." + className + ".before {\n      position: absolute;\n      top: 0;\n      left: 0;\n      transform: translateX(-100%);\n      padding-right: 4px;\n    }\n    ." + className + ".after {\n      display: inline-block;\n      padding-left: 4px;\n    }\n    h1 ." + className + " svg,\n    h2 ." + className + " svg,\n    h3 ." + className + " svg,\n    h4 ." + className + " svg,\n    h5 ." + className + " svg,\n    h6 ." + className + " svg {\n      visibility: hidden;\n    }\n    h1:hover ." + className + " svg,\n    h2:hover ." + className + " svg,\n    h3:hover ." + className + " svg,\n    h4:hover ." + className + " svg,\n    h5:hover ." + className + " svg,\n    h6:hover ." + className + " svg,\n    h1 ." + className + ":focus svg,\n    h2 ." + className + ":focus svg,\n    h3 ." + className + ":focus svg,\n    h4 ." + className + ":focus svg,\n    h5 ." + className + ":focus svg,\n    h6 ." + className + ":focus svg {\n      visibility: visible;\n    }\n  "; // This script used to have `let scrollTop` and `let clientTop` instead of
  // current ones which are `var`. It is changed due to incompatibility with
  // older browsers (that do not implement `let`). See:
  //  - https://github.com/gatsbyjs/gatsby/issues/21058
  //  - https://github.com/gatsbyjs/gatsby/pull/21083

  var script = "\n    document.addEventListener(\"DOMContentLoaded\", function(event) {\n      var hash = window.decodeURI(location.hash.replace('#', '').replace(/\u011E/g, 'g').replace(/\xDC/g, 'u').replace(/\u015E/g, 's').replace(/I/g, 'i').replace(/\u0130/g, 'i').replace(/\xD6/g, 'o').replace(/\xC7/g, 'c').replace(/\u011F/g, 'g').replace(/\xFC/g, 'u').replace(/\u015F/g, 's').replace(/\u0131/g, 'i').replace(/\xF6/g, 'o').replace(/\xE7/g, 'c').replace(/i\u0307/g, \"i\").replace(/-+/g, \"-\"))\n      if (hash !== '') {\n        var element = document.getElementById(hash)\n        if (element) {\n          var scrollTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop\n          var clientTop = document.documentElement.clientTop || document.body.clientTop || 0\n          var offset = element.getBoundingClientRect().top + scrollTop - clientTop\n          // Wait for the browser to finish rendering before scrolling.\n          setTimeout((function() {\n            window.scrollTo(0, offset - " + offsetY + ")\n          }), 0)\n        }\n      }\n    })\n  ";
  var style = icon ? /*#__PURE__*/_react.default.createElement("style", {
    key: "gatsby-remark-autolink-headers-style",
    type: "text/css"
  }, styles) : undefined;
  return setHeadComponents([style, /*#__PURE__*/_react.default.createElement("script", {
    key: "gatsby-remark-autolink-headers-script",
    dangerouslySetInnerHTML: {
      __html: script
    }
  })]);
};

gatsby-browser.js
"use strict";

var offsetY = 0;

var getTargetOffset = function getTargetOffset(hash) {
  var id = window.decodeURI(hash.replace("#", "").replace(/Ğ/g, 'g').replace(/Ü/g, 'u').replace(/Ş/g, 's').replace(/I/g, 'i').replace(/İ/g, 'i').replace(/Ö/g, 'o').replace(/Ç/g, 'c').replace(/ğ/g, 'g').replace(/ü/g, 'u').replace(/ş/g, 's').replace(/ı/g, 'i').replace(/ö/g, 'o').replace(/ç/g, 'c').replace(/i̇/g, "i").replace(/-+/g, "-"));

  if (id !== "") {
    var element = document.getElementById(id);

    if (element) {
      var scrollTop = window.pageYOffset || document.documentElement.scrollTop || document.body.scrollTop;
      var clientTop = document.documentElement.clientTop || document.body.clientTop || 0;
      var computedStyles = window.getComputedStyle(element);
      var scrollMarginTop = computedStyles.getPropertyValue("scroll-margin-top") || computedStyles.getPropertyValue("scroll-snap-margin-top") || "0px";
      return element.getBoundingClientRect().top + scrollTop - parseInt(scrollMarginTop, 10) - clientTop - offsetY;
    }
  }

  return null;
};

exports.onInitialClientRender = function (_, pluginOptions) {
  if (pluginOptions.offsetY) {
    offsetY = pluginOptions.offsetY;
  }

  requestAnimationFrame(function () {
    var offset = getTargetOffset(window.location.hash);

    if (offset !== null) {
      window.scrollTo(0, offset);
    }
  });
};

exports.shouldUpdateScroll = function (_ref) {
  var location = _ref.routerProps.location;
  var offset = getTargetOffset(location.hash);
  return offset !== null ? [0, offset] : true;
};

index.js
"use strict";

var toString = require("mdast-util-to-string");

var visit = require("unist-util-visit");

var slugs = require("github-slugger")();

var deburr = require("lodash/deburr");

function patch(context, key, value) {
  if (!context[key]) {
    context[key] = value;
  }

  return context[key];
}

var svgIcon = "<svg aria-hidden=\"true\" focusable=\"false\" height=\"16\" version=\"1.1\" viewBox=\"0 0 16 16\" width=\"16\"><path fill-rule=\"evenodd\" d=\"M4 9h1v1H4c-1.5 0-3-1.69-3-3.5S2.55 3 4 3h4c1.45 0 3 1.69 3 3.5 0 1.41-.91 2.72-2 3.25V8.59c.58-.45 1-1.27 1-2.09C10 5.22 8.98 4 8 4H4c-.98 0-2 1.22-2 2.5S3 9 4 9zm9-3h-1v1h1c1 0 2 1.22 2 2.5S13.98 12 13 12H9c-.98 0-2-1.22-2-2.5 0-.83.42-1.64 1-2.09V6.25c-1.09.53-2 1.84-2 3.25C6 11.31 7.55 13 9 13h4c1.45 0 3-1.69 3-3.5S14.5 6 13 6z\"></path></svg>";

module.exports = function (_ref, _ref2) {
  var markdownAST = _ref.markdownAST;
  var _ref2$icon = _ref2.icon,
      icon = _ref2$icon === void 0 ? svgIcon : _ref2$icon,
      _ref2$className = _ref2.className,
      className = _ref2$className === void 0 ? "anchor" : _ref2$className,
      _ref2$maintainCase = _ref2.maintainCase,
      maintainCase = _ref2$maintainCase === void 0 ? false : _ref2$maintainCase,
      _ref2$removeAccents = _ref2.removeAccents,
      removeAccents = _ref2$removeAccents === void 0 ? false : _ref2$removeAccents,
      _ref2$enableCustomId = _ref2.enableCustomId,
      enableCustomId = _ref2$enableCustomId === void 0 ? false : _ref2$enableCustomId,
      _ref2$isIconAfterHead = _ref2.isIconAfterHeader,
      isIconAfterHeader = _ref2$isIconAfterHead === void 0 ? false : _ref2$isIconAfterHead,
      _ref2$elements = _ref2.elements,
      elements = _ref2$elements === void 0 ? null : _ref2$elements;
  slugs.reset();
  visit(markdownAST, "heading", function (node) {
    // If elements array exists, do not create links for heading types not included in array
    if (Array.isArray(elements) && !elements.includes("h" + node.depth)) {
      return;
    }

    var id;

    if (enableCustomId && node.children.length > 0) {
      var last = node.children[node.children.length - 1]; // This regex matches to preceding spaces and {#custom-id} at the end of a string.
      // Also, checks the text of node won't be empty after the removal of {#custom-id}.

      var match = /^(.*?)\s*\{#([\w-]+)\}$/.exec(toString(last));

      if (match && (match[1] || node.children.length > 1)) {
        id = match[2]; // Remove the custom ID from the original text.

        if (match[1]) {
          last.value = match[1];
        } else {
          node.children.pop();
        }
      }
    }

    if (!id) {
      var slug = slugs.slug(toString(node), maintainCase);
      id = removeAccents ? deburr(slug) : slug;
    }

    var data = patch(node, "data", {});
    id = id.replace(/Ğ/g, 'g').replace(/Ü/g, 'u').replace(/Ş/g, 's').replace(/I/g, 'i').replace(/İ/g, 'i').replace(/Ö/g, 'o').replace(/Ç/g, 'c').replace(/ğ/g, 'g').replace(/ü/g, 'u').replace(/ş/g, 's').replace(/ı/g, 'i').replace(/ö/g, 'o').replace(/ç/g, 'c').replace(/i̇/g, "i").replace(/-+/g, "-");
    patch(data, "id", id);
    patch(data, "htmlAttributes", {});
    patch(data, "hProperties", {});
    patch(data.htmlAttributes, "id", id);
    patch(data.hProperties, "id", id);

    if (icon !== false) {
      patch(data.hProperties, "style", "position:relative;");
      var label = id.split("-").join(" ");
      var method = isIconAfterHeader ? "push" : "unshift";
      node.children[method]({
        type: "link",
        url: "#" + id,
        title: null,
        children: [],
        data: {
          hProperties: {
            "aria-label": label + " permalink",
            class: className + " " + (isIconAfterHeader ? "after" : "before")
          },
          hChildren: [{
            type: "raw",
            // The Octicon link icon is the default. But users can set their own icon via the "icon" option.
            value: icon
          }]
        }
      });
    }
  });
  return markdownAST;
};